### DISCUSSION TITLE
Help understanding Falcon/Rails/ActiveRecord Limitations and Future

### DISCUSSION THREAD

**Discussion by tleish** [socketry/falcon#186] [Q&A]
## QUESTIONS:

- Is Rails on Falcon a viable production solution?
- What are the gotchas with using Falcon and Rails (specifically with ActiveRecord)
- What is the future for Falcon?

## DETAILS:

A recent article on ideas for improving the Ruby JIT compiler sparked a conversation between devs in [Hackernews](https://news.ycombinator.com/item?id=32408716) discussing different approaches to making a Ruby on Rails application run faster.  If a Rails app is CPU bound, then improving the JIT might help.  However, if the Rails app is primarily I/O bound, then this approach might not make much of a difference.

A former Shopify developer wrote:

> I worked at Shopify for several years. The big problems were *always* due to I/O. Usually that was in the form of long-running network calls. You're calling the FedEx API for shipping rates in the cart. All of a sudden FedEx starts taking 5 seconds instead of 1 second, and all of your worker capacity evaporates. That kind of thing.
>
> [Hacker News response](https://news.ycombinator.com/item?id=32411553)

We avoid I/O delay by deploying two servers.  A Rails server and an asynchronous Ruby proxy server (falcon).

### Standard Request

An example of a standard internal request looks like this.

```mermaid
sequenceDiagram
    Browser->>Rails: Submit Form
    activate Rails
    Rails->>MySQL: Query User Info
    activate MySQL
    MySQL-->>Rails: User Info
    deactivate MySQL
    
    Rails->>MySQL: Update User Info
    activate MySQL
    MySQL-->>Rails: Update Status
    deactivate MySQL
    deactivate Rails

    Rails-->>Browser: Response
```

### 3rd-Party Request

If a user interaction requires I/O delay from a 3rd-party API, the sequence looks like this.

```mermaid
sequenceDiagram
    Browser->>Falcon: Submit Form
    activate Falcon

    Falcon->>Rails: Get User Info (validate user)
    activate Rails
    Rails->>MySQL: Query User Info
    activate MySQL
    MySQL-->>Rails: User Info
    deactivate MySQL
    Rails-->>Falcon: Return User Info
    deactivate Rails
    
    Falcon->>3rd Party API: Submit Form w/ User Info
    activate 3rd Party API
    3rd Party API-->>Falcon: API Response
    deactivate 3rd Party API
       
    Falcon->>Rails: Update Record
    activate Rails
    Rails->>MySQL: Update User Info
    activate MySQL
    MySQL-->>Rails: Update Status
    deactivate MySQL
    Rails-->>Falcon: Response
    deactivate Rails

    Falcon-->>Browser: Response
    deactivate Falcon
```

Using Falcon as an asynchronous proxy prevents our Rails server resources from getting tied up while waiting for 3rd party I/O.  For us, this has scaled very well.  As you can see, the main downside is the additional complication.  I'd love to combine Falcon/Rails into a single server to simplify this.  However, it's unclear to me the current status/roadmap for making this work, especially for a legacy app.

In researching this I've found the following:

> The only caveat is that it doesn't work with Ruby on Rails, because `ActiveRecord` doesn't support `async` gem. You can still use it with Rails if `ActiveRecord` is not involved.
>
> See: [Async Ruby - Bruno Sutic](https://brunosutic.com/blog/async-ruby)

In response to the above article, @ioquatix responded:

> ActiveRecord does work but it’s hugely limited because they have explicit per-thread resource pools which we can’t get around very easily.
>
> See:  [Hackernews discussion](https://news.ycombinator.com/item?id=29050482) 

What does "hugely limited" mean?  Does it just mean that ActiveRecord would be synchronous, while a 3rd party request could be asynchronous and release the fiber while waiting for response?  This is how our stack works today, but this would just mean we could consolidate to a single server.

```mermaid
sequenceDiagram
    Browser->>Falcon/Rails: Submit Form
    activate Falcon/Rails

    Falcon/Rails->>MySQL: Query User Info <br>(use fiber resource)
    activate MySQL
    MySQL-->>Falcon/Rails: User Info
    deactivate MySQL

    
    Falcon/Rails->>3rd Party API: Submit Form w/ User Info <br>(wait, release fiber resource)
    activate 3rd Party API
    3rd Party API-->>Falcon/Rails: API Response
    deactivate 3rd Party API
       
    Falcon/Rails->>MySQL: Update User Info <br>(use fiber resource)
    activate MySQL
    MySQL-->>Falcon/Rails: Update Status
    deactivate MySQL

    Falcon/Rails-->>Browser: Response
    deactivate Falcon/Rails
```

I also found [Falcon Guides > Rails Integration](https://socketry.github.io/falcon/guides/rails-integration/index.html) which does not mention any gotchas or limitations of ActiveRecord.

I know there's been some recent changes to Ruby 3 and fibers, is it advised to wait for these to be deployed?

In my research I also remember reading a comment by @ioquatix suggesting Falcon as experimental and not to be used for production (which we currently are).

**Comment by ioquatix** [comment]
Wow that's an awesome summary with really great diagrams. My main focus right now is Rack 3, after that I'll move on to Rails and we will solve any of the issues that still exist there.

The recent work with https://github.com/rails/rails/pull/43596 will help set the stage for connection per fiber.

Right now with AR, the worst case is you will have one connection per thread which means every incoming request (fiber) on the thread would contend with each other (over a single AR connection). While it works, if you are database bound, it is less than ideal. That's based on last time I checked the state of Rails, but it has changed, I don't know if #43596 made it into Rails 7 - and I haven't evaluated the impact of it yet.

There is also work going on in the database backends, e.g. `pg` https://github.com/ged/ruby-pg/pull/397

The MySQL gem is a little harder because the actual C interface is insufficient, but `libmariadb` is good enough and I have shown that in https://github.com/socketry/db-mariadb - I would suggest one option would be to use the `db-mariadb` gem either directly or as a driver behind ActiveRecord.

**Reply by tleish** [reply]
> Right now with AR, the worst case is you will have one connection per thread which means every incoming request (fiber) on the thread would contend with each other (over a single AR connection). While it works, if you are database bound, it is less than ideal. That's based on last time I checked the state of Rails, but it has changed, I don't know if #43596 made it into Rails 7 - and I haven't evaluated the impact of it yet.

Can you clarify this.  So, in the last diagram shared above which of the following is true?  
- The entire request (end-to-end) consumes a fiber when querying DB and waiting for 3rd-party API request response
- Part of the request consumes a fiber when querying DB, but not when waiting for 3rd-party API request response

**Reply by ioquatix** [reply]
Last time I checked this, it worked like this:

- ActiveRecord will checkout a single connection per thread, so Falcon with 8 threads will have 8 connections, one per thread.
- Falcon will load balance requests. Assuming each request does a slow DB query, and you have 8 threads, if you have 9 requests, 1 thread will be processing two requests.
- Of those two requests, both are performing a slow SQL query. IIRC, each ActiveRecord connection has a mutex, which [IIUC is locked for each query/transaction](https://github.com/rails/rails/blob/d6973d50d01fe55e5c65368b92160708fb60a3da/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb#L908).
- This will serialise those two requests on that thread.

I'll follow up with some specific benchmarks.

**Reply by ioquatix** [reply]
There is a benchmark for comparing these different scenarios here: https://github.com/socketry/db/tree/main/benchmark/compare which does `pg_sleep(1)` for several different scenarios. I think some parts of this benchmark have accrued some bit-rot, but it does appear to show the key differences.

## `db` gem

Using the `db` gem, we expect this to scale fairly linearly until `pg` starts dying.

```
> benchmark-http concurrency https://localhost:9292/db
I am going to benchmark https://localhost:9292/db...

I am running 1 asynchronous tasks that will each make sequential requests...
I made 11 requests in 11.0s. The per-request latency was 1.0s. That's 1.0 asynchronous requests/second.
	          Variance: 1.199µs
	Standard Deviation: 1.09ms
	    Standard Error: 0.00033010285588397646

I am running 2 asynchronous tasks that will each make sequential requests...
I made 22 requests in 11.1s. The per-request latency was 1.0s. That's 1.99 asynchronous requests/second.
	          Variance: 15.311µs
	Standard Deviation: 3.91ms
	    Standard Error: 0.000834251153302291

I am running 4 asynchronous tasks that will each make sequential requests...
I made 44 requests in 11.0s. The per-request latency was 1.0s. That's 3.99 asynchronous requests/second.
	          Variance: 0.410µs
	Standard Deviation: 640.594µs
	    Standard Error: 9.657320275148998e-05

I am running 8 asynchronous tasks that will each make sequential requests...
I made 88 requests in 11.0s. The per-request latency was 1.0s. That's 7.98 asynchronous requests/second.
	          Variance: 1.586µs
	Standard Deviation: 1.26ms
	    Standard Error: 0.0001342576957067329

I am running 16 asynchronous tasks that will each make sequential requests...
I made 176 requests in 11.0s. The per-request latency was 1.0s. That's 15.96 asynchronous requests/second.
	          Variance: 0.974µs
	Standard Deviation: 987.023µs
	    Standard Error: 7.43996436975189e-05

I am running 32 asynchronous tasks that will each make sequential requests...
I made 352 requests in 11.0s. The per-request latency was 1.0s. That's 31.94 asynchronous requests/second.
	          Variance: 0.190µs
	Standard Deviation: 436.029µs
	    Standard Error: 2.3240397599205717e-05

I am running 64 asynchronous tasks that will each make sequential requests...
I made 704 requests in 11.0s. The per-request latency was 1.0s. That's 63.83 asynchronous requests/second.
	          Variance: 5.991µs
	Standard Deviation: 2.45ms
	    Standard Error: 9.225139365665314e-05
```

## `ActiveRecord::Base.connection.execute`

```
> benchmark-http concurrency https://localhost:9292/active_record
I am going to benchmark https://localhost:9292/active_record...

I am running 1 asynchronous tasks that will each make sequential requests...
I made 11 requests in 11.0s. The per-request latency was 1.0s. That's 1.0 asynchronous requests/second.
	          Variance: 11.611µs
	Standard Deviation: 3.41ms
	    Standard Error: 0.0010273808557839084

I am running 2 asynchronous tasks that will each make sequential requests...
I made 51 requests in 50.6s. The per-request latency was 2.0s. That's 1.01 asynchronous requests/second.
	          Variance: 19.27ms
	Standard Deviation: 138.81ms
	    Standard Error: 0.019437566944935123
Your server can handle 1 concurrent requests.
At this level of concurrency, requests have ~1.0x higher latency.
```

We can get into more details, but this is effectively the issue I'm talking about. It might be nice to test the new per-fiber execution state to see if we can improve the situation.

**Reply by tleish** [reply]
Thanks for taking the time to explain and illustrate the bottleneck currently in ActiveRecord.

As I mentioned in my original post, we run a Falcon (proxy server) for communicating with external I/O and Rails on Puma for internal (ActiveRecord) I/O.  It's clear that if I run ActiveRecord on Falcon, I would gain anything.  However, I'd like to understand what happens when I run both internal (ActiveRecord) and external I/O on a single Rails server?  In other words a single Rails request first communicates with an external API and handles an ActiveRecord query (or the reverse).  

|                                           | Option 1: Falcon Proxy |            |
| ----------------------------------------- | ------------ | ------------------------- |
| Step 1: Browser Sends Request             | Falcon Proxy | **Fiber**     |
| Step 2: Query DB w/ActiveRecord           | Puma/Rails   | Thread        |
| Step 3: Communicate with external API | Falcon Proxy | **Fiber** |
| Step 4: Save to DB w/ActiveRecord         | Puma/Rails   | Thread        |
|  |  |  |
|  | **Option 2: Falcon (No Proxy)** |  |
| **Step 1: Browser Sends Request** | Falcon/Rails | **Fiber or Thread?** |
| Step 2: Query DB w/ActiveRecord | Falcon/Rails | Thread |
| **Step 3: Communicate with external API** | Falcon/Rails | **Fiber or Thread?** |
| Step 4: Save to DB w/ActiveRecord | Falcon/Rails | Thread |
|  |  | Potential same Latency as proxy or worse? |

Will either setup use a Fiber/Thread/Fiber/Thread pattern, or is the second setup all Thread?  In other words, while the ActiveRecord query still consumes 1 thread, when the request moves on to the external request does it use an async fiber and you see the same performance as if the 3rd party request was on a proxy server?  Or, does ActiveRecord somehow tie up the entire end-to-end request as a thread and does not use any fibers?

**Reply by tleish** [reply]
Annoying how github discussions hides additional comments of an answer behind a "show more" link.

**Reply by trevorturk** [reply]
Agreed. Not sure why they did that. Same with long comment threads, too, imo :/

I think you can wait for Samuel to chime in, but also might be worth testing this yourself. I think you could use the benchmark stuff mentioned above (https://github.com/socketry/db/tree/main/benchmark/compare) -- I suppose you'd just add some non-database work the happens on either side of the database work to see if that work scales up linearly.

Looking at the benchmark and trying to suss it out, my guess for the "does ActiveRecord somehow tie up the entire end-to-end request as a thread and does not use any fibers" question is "no" -- rather, ActiveRecord will use a single connection, so any single ActiveRecord query will block any other ActiveRecord query as well, but not necessarily all other work. All ActiveRecord work would be waiting on a single connection. So, if you have super fast queries, oh well, but if you have some slow queries it could be bad, because only one ActiveRecord query could happen at once, so the other fibers will be held up waiting for the single connection to be freed up. So, I suppose it's a bit more complicated than a simple "no" but the practical impact would depend on your specific app. 

In my case, I'm thinking of either using the db gem until the ActiveRecord connection pool stuff is sorted out, or using caching around everything. That being said, to complicate things further, I'm using Rails.cache with a Redis backend, and I'm really not sure if that suffers from the same issue as ActiveRecord here, where everything would be sharing a single connection. My Redis work is so quick, though, if I am encountering a bottleneck it hasn't been severe enough to be noticed. Similar to the db gem, perhaps I should be using https://github.com/socketry/async-redis for now, or I should figure out how to run the benchmark with a Redis sleep or w/e to prove if we have the same issue or not. I confess things are working well enough for me (so far) that I haven't dug into this too deeply (yet).

**Reply by ioquatix** [reply]
@trevorturk is largely correct, the contention is only with the single ActiveRecord connection and so you'd be fine to use Rails provided you don't contend on the connection. Option 2 should be fine, and will only get better as we wrap up some of the downstream implementation details.

IF you use Ruby 3.1+ (and disable zeitwerk/autoload because it's buggy in 3.1.x, fixed in 3.2), you can wrap your database queries in `Thread.new{... Model.find_by(...) ...}.value`. This will be less efficient and less elegant but it will avoid the contention since thread pool will work correctly. `Thread#join` is non-blocking on Ruby 3.1+. Bearing in mind, the thread poll has a limited number of connections too.

**Reply by trevorturk** [reply]
Thanks for the `Thread.new...` hack idea, that should help quite a bit!

I wanted to note, I think (?) it's fair to handle the "disable zeitwerk/autoload" gotcha by using Zeitwerk's global eager load: https://github.com/fxn/zeitwerk#global-eager-load -- it also looks like there's a "Rails way" to do this as well, documented here: https://guides.rubyonrails.org/classic_to_zeitwerk_howto.html#check-zeitwerk-compliance-in-the-test-suite -- this is written as though you'd only use it for testing purposes, but if autoload is buggy in 3.1.x, doing the global eager load either way should avoid the bugs. I'm doing this in an initializer in my Falcon/Rails 3.1.x app now on @ioquatix's advice.

**Comment by trevorturk** [comment]
Thanks for this great discussion. I wanted to chime in to say that I'm very happily using Rails + Falcon in production, but without a database. IIRC @ioquatix had suggested considering https://github.com/socketry/db but I understand that some work has gone into making ActiveRecord work better when running with Falcon, and sounds like more is coming, which is great news. 

I wanted to point out also some helpful discussion here: https://github.com/rails/rails/pull/44219#issuecomment-1050812033

I'm hoping @ioquatix can clarify the "entire request" vs "part of the request" question, but my (naive) understanding is that ActiveRecord would be operating a single connection, as opposed to a connection pool. If that's the case, it's not really a question of "consuming a fiber" vs not using a connection pool. 

In my case, I'm just about to add a database to my app, but I was thinking of using `Rails.cache` around the read-only data needed to serve a request, and then enqueueing a Sidekiq job in an `after_action` callback to write to the database. (I have a sort of strange use-case that makes this a tenable solution for my short term needs, but I'm very excited for future improvements to ActiveRecord when used with Falcon/Fibers.)

**Reply by tleish** [reply]
Thank you.  The discussion at https://github.com/rails/rails/pull/44219#issuecomment-1050812033 was helpful in understanding all the Rails components tied to threads and how big the task is to convert Rails to be fiber friendly.

**Comment by j-manu** [comment]
@tleish Can you provide some more details about the asynchronous Ruby proxy server?

1. Is it using any framework?
2. Does it proxy all routes or selected routes? 
3. Are the calls from proxy server to Rails app through HTTP? If so, how is auth handled?

Any code samples you can share will be of great help.

**Comment by tleish** [comment]
@j-manu

**_Q: Is it using any framework?_**
A: The proxy stack is [grape](https://github.com/ruby-grape/grape) on falcon

**_Q: Does it proxy all routes or selected routes?_**
A: The browser only sends request to the falcon proxy which in involve 3rd-party requests (this is quite a few for us), all other requests are sent directly to the Rails server. A load balancer delegates all requests with `/api` to the falcon server, and then the rest to the Rails server.  

**_Q: Are the calls from proxy server to Rails app through HTTP? If so, how is auth handled?_**
A: With both the Falcon and Rails behind a load balancer, they both share the same domain.  Sharing the same domain means they can  share cookies.  The falcon server relays the cookie in the _https_ request to the Rails server.  Nothing special setup on the Rails server to handle the requests from the API as they just look like a direct browser request.

Both servers can access Redis (sessions), but only the Rails server accesses the DB.  Rails server never sends requests to Falcon or 3rd Party server to maintain low latency.

![image](https://user-images.githubusercontent.com/1099993/186700975-d59fa871-a1e8-4add-a79f-bc14c55962d1.png)

**Reply by j-manu** [reply]
@tleish This has been of great help. Thank you.

**Comment by nic-lan** [comment]
👋 

I would like to understand if there is any update.
I understand that because of https://github.com/rails/rails/pull/44219 AR ConnectionPool is now Fiber-safe.

**Comment by sdhull** [comment]
Soooo I just replaced `puma` with `falcon` in my rails app and it seems like it "Just works." Running `rails s` now gives:
```
=> Booting Falcon
=> Rails 7.0.5 application starting in development http://localhost:3000
=> Run `bin/rails server --help` for more startup options
```
I'm stoked! But also I have.... questions:
1. How the heck does this work?
2. Is there anything else I should do? I remember back in the EventMachine / EM-Synchrony days there were a bunch of `em-` prefixed gems to make things "fiber-aware"
3. Is there an easy way to add configuration for falcon? There's a `config/puma.rb` file for configuring puma... is there anything like this for falcon? Also, does anyone know if running `rails serve` is more like `falcon serve` or more like `falcon host`?
4. This brings me to my next question: docs say `falcon serve` is not intended for deployment, and you should use `falcon host` instead. Can you elaborate a bit? _Why_ isn't `falcon serve` intended for deployment? What would happen if I run my production server like that?
5. `falcon host` does not seem to work (in my rails app) with either of the example `falcon.rb` file contents, any idea what's going on here?
```
❯ falcon host
  0.0s     info: Falcon::Command::Host [oid=0xa28] [ec=0xa3c] [pid=66766] [2023-07-21 00:40:13 -0700]
               | Falcon Host v0.42.3 taking flight!
               | - Configuration: falcon.rb
               | - To terminate: Ctrl-C or kill 66766
               | - To reload: kill -HUP 66766
<internal:/Users/stevehull/.rbenv/versions/3.2.2/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:37:in `require': cannot load such file -- process/metrics (LoadError)
	from <internal:/Users/stevehull/.rbenv/versions/3.2.2/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:37:in `require'
	from /Users/stevehull/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/falcon-0.42.3/lib/falcon/service/supervisor.rb:23:in `<top (required)>'
	from /Users/stevehull/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/falcon-0.42.3/lib/falcon/environments/supervisor.rb:23:in `require_relative'
	from /Users/stevehull/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/falcon-0.42.3/lib/falcon/environments/supervisor.rb:23:in `block in load'
	from /Users/stevehull/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/falcon-0.42.3/lib/falcon/configuration.rb:134:in `instance_eval'
	from /Users/stevehull/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/falcon-0.42.3/lib/falcon/configuration.rb:134:in `block in load'
	from /Users/stevehull/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/falcon-0.42.3/lib/falcon/configuration.rb:127:in `each'
	from /Users/stevehull/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/falcon-0.42.3/lib/falcon/configuration.rb:127:in `load'
	from /Users/stevehull/Dev/radix-on-rails/falcon.rb:4:in `load_file'
	from /Users/stevehull/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/falcon-0.42.3/lib/falcon/configuration.rb:118:in `instance_eval'
	from /Users/stevehull/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/falcon-0.42.3/lib/falcon/configuration.rb:118:in `load_file'
	from /Users/stevehull/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/falcon-0.42.3/lib/falcon/configuration.rb:85:in `load_file'
	from /Users/stevehull/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/falcon-0.42.3/lib/falcon/command/host.rb:54:in `block in configuration'
	from /Users/stevehull/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/falcon-0.42.3/lib/falcon/command/host.rb:52:in `each'
	from /Users/stevehull/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/falcon-0.42.3/lib/falcon/command/host.rb:52:in `configuration'
	from /Users/stevehull/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/falcon-0.42.3/lib/falcon/controller/host.rb:39:in `initialize'
	from /Users/stevehull/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/falcon-0.42.3/lib/falcon/command/host.rb:62:in `new'
	from /Users/stevehull/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/falcon-0.42.3/lib/falcon/command/host.rb:62:in `controller'
	from /Users/stevehull/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/falcon-0.42.3/lib/falcon/command/host.rb:84:in `call'
	from /Users/stevehull/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/falcon-0.42.3/lib/falcon/command/top.rb:105:in `call'
	from /Users/stevehull/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/samovar-2.2.0/lib/samovar/command.rb:21:in `call'
	from /Users/stevehull/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/falcon-0.42.3/lib/falcon/command.rb:30:in `call'
	from /Users/stevehull/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/falcon-0.42.3/bin/falcon:26:in `<top (required)>'
	from /Users/stevehull/.rbenv/versions/3.2.2/bin/falcon:25:in `load'
	from /Users/stevehull/.rbenv/versions/3.2.2/bin/falcon:25:in `<main>'
```

**Reply by sdhull** [reply]
Great news! I found [the rails example](https://github.com/socketry/rails-falcon-heroku) and copied the `falcon.rb` file there (with slight variations) + the `preload.rb` file and voila! `bundle exec falcon host` now works!

I'm going to forge ahead. I'm already using `async/barrier` to fire off 20 requests concurrently for a certain endpoint. Reduced latency there from ~20 seconds to ~1 second, so already a nice win.

**Reply by ioquatix** [reply]
ActiveRecord won't be fully compatible until Rails 7.1 is released. However, IIRC, it's not totally broken, it's just essentially limited to "1 connection per thread", which since Falcon only uses a single thread, is a bit limiting. You could try using rails head if you want to forge ahead with something. I believe it all works quite okay.

**Reply by trevorturk** [reply]
Yes I’m using Rails 7.0 and it works well, with the connection limitation caveat mentioned. 

Rails 7.1 isolation_state config should help, but if you search the Rails issues you’ll come across some other relatively minor inefficiencies that may want to be addressed eventually.

**Reply by sdhull** [reply]
> with the connection limitation

What's the upshot of this? Is it essentially that only one request can be served concurrently? Does falcon offer forking or threading to mix concurrency models at all?

**Reply by trevorturk** [reply]
I run one Falcon worker per Heroku Dyno so it means one connection is shared between all requests on the same Dyno. Since I’m using the pg gem which is fiber-aware it basically seems fine, but I’m sure it’ll be optimized better when switch to isolation_level :fiber in Rails 7.1. 

I’d note though that I’m a bit nervous about swapping things in production since I believe it’ll open a connection per fiber, which can be a lot of connections. I believe Samuel’s advice is to create your own pool using an Async Semaphore with a limit, but I haven’t gotten that far yet. When we get there let’s all agree to pitch in with docs and best practices please! :)

**Comment by sdhull** [comment]
FWIW I was running falcon on fly.io and I noticed 2 things:
- memory usage was nearly 2x compared to puma (~450MB vs ~250MB)
- as soon as concurrent users started going up, people were getting 429's. resolved with puma.

I can't help but wonder if a forking + evented model like rainbows would be worth trying. not sure what to make of the memory footprint

**Reply by trevorturk** [reply]
Interesting! FWIW I'm using Falcon on Heroku and it's been great, much more efficient than Puma, because I'm waiting on a lot of upstream API responses, Fibers are a lighter solution than threads, allowing me to handle more concurrent requests with fewer dynos.

Re: memory usage, did you tweak the `count` to match your number of CPUs? My config looks like this: https://gist.github.com/trevorturk/0ef612e365e033549d3a30aa5947ae5c for Standard 1X dynos. 

Re: 429s, I'm not sure if that would be Falcon -- wonder if it could be fly.io limiting the number of concurrent requests per dyno (or whatever their "dyno" equivalent thing is called)

**Reply by ioquatix** [reply]
@sdhull how many processors are available on Fly.io? At a guess, you could be comparing multi-process falcon with multi-thread puma. If you didn't use `preload.rb`, falcon may use more memory.

Not sure about 429s, never seen that before, but I've definitely seen issues with proxies in front of Falcon being unable to handle the load correctly. I don't think Falcon itself can return 429s :)

It would be nice to understand a bit more about the environment.

**Reply by taltas** [reply]
@sdhull Hi, I'm interested to know if you ended up resolving your memory issues with Falcon? Did you put it back in or did you go back to sticking with Puma?

**Comment by osbre** [comment]
Hi guys! I quickly went through this conversion and would love to learn more. Here's what I got so far:

1. Falcon is awesome because it uses Fibers for all incoming requests, makes things like external HTTP calls non-blocking and thus increases the throughput of a plain Rails application by a lot.

2. ActiveRecord cannot benefit from Fibers because it's powered by a blocking DB driver. There would need to be written a bridge / special driver for it that utilizes [socketry/db](https://github.com/socketry/db).

If I got it correctly, could you tell me more info as to what makes it so hard to make the ActiveRecord compatible? Would love to learn about this complexity and fill the gaps in my head.

Also found this repo: https://github.com/socketry/db-active_record. Does it imply that such driver is possible but requires a lot of testing and effort?

**Reply by ioquatix** [reply]
For (2), the `pg` gem, and thus the Postgres driver for ActiveRecord, is compatible with Async/event driven I/O. However, the `db` gem takes it a step further and utilises the async capabilities of `libpq` and `libmariadbclient`, designed from the ground up to be concurrent. I would like to work on `db-active_record` more, but it's a huge effort. We need to (1) improve the Rails ActiveRecord interfaces to make them more generic, and (2) implement those interfaces fully in `db-active_record`.

**Reply by osbre** [reply]
@ioquatix It looks like Rage, a somewhat Rails-compatible framework using Fibres, has added support for ActiveRecord.
https://github.com/rage-rb/rage/pull/103
Sharing this in case it helps
